File: WangLulu/Trie/ReplaceWords.java
package Trie;

import java.util.List;

public class ReplaceWords {
    public String replaceWords(List<String> dictionary, String sentence) {
        Trie trie = new Trie();
        for(String root:dictionary){
            trie.insert(root);
        }
        String[] words = sentence.split(" ");
        for(int i = 0 ;i<words.length;i++){
            String shortestRoot = trie.find(words[i]);
            if(shortestRoot!=null){
                words[i] = shortestRoot;
            }
        }
        return String.join(" ",words);
    }
}
class TrieNode{
    TrieNode[] children = new TrieNode[26];
    boolean isRoot;
    public TrieNode(){
        isRoot = false;
    }
}
class Trie{
    private TrieNode root;
    public Trie(){
        root = new TrieNode();
    }
    public void insert(String rootword){
        TrieNode cur = root;
        for(char c :rootword.toCharArray()){
            int index = c -'a';
            if(cur.children[index] == null){
                cur.children[index] = new TrieNode();
            }
            cur = cur.children[index];
        }
        cur.isRoot = true;
    }
    public String find(String words){
        TrieNode cur = root;
        StringBuilder sb = new StringBuilder();
        for(char c :words.toCharArray()){
            int index = c - 'a';
            if(cur.children[index] == null){
                break;
            }
            cur = cur.children[index];
            sb.append(c);
            if(cur.isRoot){
                return sb.toString();
            }
        }
        return null;
    }
}

File: ShengJie Wang/Practice_3/ReplaceWords.cpp
/* Replace Words */

struct TrieNode{
    TrieNode* nexts[26];
    int endby;
    TrieNode(){
        fill(nexts, nexts + 26, nullptr);
        endby = 0;
    }
};

class Trie{
public:
    Trie(){
        root = new TrieNode();
    }
    
    void Insert(string word){
        TrieNode* cur = root;
        for(char c : word){
            int index = c - 'a';
            if(!cur->nexts[index])
                cur->nexts[index] = new TrieNode();
            cur = cur->nexts[index];
        }
        cur->endby = true;
    }
    
    string find(string word){
        string ans;
        TrieNode* cur = root;
        for(char c : word){
            int index = c - 'a';
            if(!cur->nexts[index])
                return word;
            ans += c;
            cur = cur->nexts[index];
            if(cur->endby)
                return ans;
        }
        return word;
    }
    TrieNode* root;
};

class Solution {
public:
    string replaceWords(vector<string>& dictionary, string sentence) {
        Trie Dictionary;
        string ans;
        for(string word : dictionary){
            Dictionary.Insert(word);
        }
        TrieNode* cur = Dictionary.root;
        string word;
        for(char c : sentence){
            if(c == ' '){
                if(!word.empty())
                    ans += Dictionary.find(word) + ' ';
                word.clear();
                continue;
            }
            word += c;
        }
        if(!word.empty())
            ans += Dictionary.find(word);
        return ans;
    }
};